Articles/FileSystemProviders/extending.md

```diff
@@ -0,0 +1,213 @@
+
+As mentioned netcore by default store static files, such as media and css, in the wwwroot folder, but we can register an additional location in the configure section of our startup
+
+In the `configure` method in startup.cs register a new static file location like so:
+
+```c#
+public void Configure(IApplicationBuilder app)
+{
+    ...
+
+    app.UseStaticFiles(new StaticFileOptions
+    {
+        FileProvider = new PhysicalFileProvider(Path.Combine("D:", "storage", "umbracoMedia")),
+        RequestPath = "/CustomPath"
+    });
+}
+```
+The PhysicalFileProvider takes a single parameter, the **RootPath** - the rooted, filesystem path, using directory separator chars, NOT ending with a separator // eg "c:" or "c:\path\to\site" or "\\server\path", the safest way to achieve this is using `Path.Combine`.
+
+
+You also have to specify the  **RequestPath** - the relative url, where the media will be served, using url separator chars, NOT ending with a separator / eg "" or "/Views" or "/Media".
+
+Now you can use your newly registered static file location as if it was wwwroot. Notice how you no longer need to use `hostingEnvironment.MapPathWebRoot(folderLocation)`, since you're no longer trying to map the location to somewhere within wwwroot, but instead use your newly registered static file location.
+
+```csharp
+public void Compose(IUmbracoBuilder builder)
+{
+    builder.SetMediaFileSystem((factory) =>
+    {
+        IHostingEnvironment hostingEnvironment = factory.GetRequiredService<IHostingEnvironment>();
+        var rootPath = Path.Combine("C:", "storage", "umbracoMedia");
+        var rootUrl = hostingEnvironment.ToAbsolute("/CustomPath");
+
+        return new PhysicalFileSystem(
+            factory.GetRequiredService<IIOHelper>(),
+            hostingEnvironment,
+            factory.GetRequiredService<ILogger<PhysicalFileSystem>>(),
+            rootPath,
+            rootUrl);
+    });
+}
+```
+
+This is almost the same as when registering a location within the wwwroot folder, the only difference is that rootPath is now set to the path we gave the `PhysicalFileProvider` and the rootUrl is the same as we set as the `RequestPath` in the `StaticFileOption`.
+
+Our media is now stored in C:\storage\umbracoMedia, and is served from the base URL /CustomPath, so an image URL will look something like mysite.com/CustomPath/MyAwesomePicture.png
+
+### IFileSystem
+
+`PhysicalFileSystem` implements the `IFileSystem` interface, and it is possible to replace it with a custom class - eg. if you want your media files stored on Azure or something similar. You replace the media filesystem using the `SetMediafileSystem` method in a composer like shown in the MediaFileSystem section, but instead of returning a PhysicalFileSystem you return whatever implementation of IFileSystem you want.
```

Comment:
```suggestion
`PhysicalFileSystem` implements the `IFileSystem` interface, and it is possible to replace it with a custom class - eg. if you want your media files stored on Azure or something similar. You replace the media filesystem using the `SetMediafileSystem` method in a composer like shown in the `MediaFileSystem` section, but instead of returning a `PhysicalFileSystem`, you return whatever implementation of `IFileSystem` you want.
```